File: sdks/local-rp/csharp/src/LinkKeys.LocalRp/Rpc/RpcClient.cs
using LinkKeys.LocalRp.Dns;
using LinkKeys.LocalRp.Wire;
using static LinkKeys.LocalRp.Wire.Types;

namespace LinkKeys.LocalRp.Rpc;

/// <summary>
/// CSIL-RPC over the injected <see cref="ITransport"/>, TLS-pinned to a domain's DNS
/// <c>fp=</c> records — this SDK's only network surface, per the design doc's "Required
/// Network Access": domain public keys, revocations, and claim-ticket redemption, all
/// unauthenticated-TLS TCP CSIL-RPC calls.
/// </summary>
public static class RpcClient
{
    /// <summary>A discovered endpoint for a domain: its pinned trust-anchor fingerprints and its CSIL-RPC TCP address.</summary>
    public sealed record DomainEndpoint(IReadOnlyList<string> Fingerprints, string TcpAddr);

    /// <summary>
    /// Look up a domain's trust anchor + TCP endpoint over DNS TXT. Fails closed: a
    /// missing/unparseable record, or a <c>_linkkeys</c> record with no <c>fp=</c>
    /// entries, or a <c>_linkkeys_apis</c> record with no <c>tcp=</c> entry, is an error.
    /// </summary>
    public static DomainEndpoint DiscoverDomainEndpoint(IDnsResolver dns, string domain)
    {
        var anchorName = LinkKeys.LocalRp.Dns.Dns.LinkKeysDnsName(domain);
        var anchorTxts = dns.TxtLookup(anchorName);
        IReadOnlyList<string>? fingerprints = null;
        foreach (var txt in anchorTxts)
        {
            try
            {
                var rec = LinkKeys.LocalRp.Dns.Dns.ParseLinkKeysTxt(txt);
                if (rec.Fingerprints.Count > 0)
                {
                    fingerprints = rec.Fingerprints;
                    break;
                }
            }
            catch (DnsParseError)
            {
                // try the next TXT record
            }
        }

        if (fingerprints is null)
        {
            throw new SdkException(SdkException.ErrorKind.Dns, $"no usable {anchorName} TXT record with fp= entries");
        }

        var apisName = LinkKeys.LocalRp.Dns.Dns.LinkKeysApisDnsName(domain);
        var apisTxts = dns.TxtLookup(apisName);
        string? tcpAddr = null;
        foreach (var txt in apisTxts)
        {
            try
            {
                var apis = LinkKeys.LocalRp.Dns.Dns.ParseLinkKeysApisTxt(txt);
                if (apis.Tcp is not null)
                {
                    tcpAddr = apis.Tcp;
                    break;
                }
            }
            catch (DnsParseError)
            {
                // try the next TXT record
            }
        }

        if (tcpAddr is null)
        {
            throw new SdkException(SdkException.ErrorKind.Dns, $"no usable {apisName} TXT record with tcp= entry");
        }

        return new DomainEndpoint(fingerprints, tcpAddr);
    }

    private static string ExtractHostname(string hostPort)
    {
        if (hostPort.StartsWith('['))
        {
            int end = hostPort.IndexOf(']');
            if (end != -1)
            {
                return hostPort[1..end];
            }
        }

        int idx = hostPort.LastIndexOf(':');
        return idx == -1 ? hostPort : hostPort[..idx];
    }

    /// <summary>Send one CSIL-RPC request over a fresh TLS connection to <paramref name="endpoint"/> and return the decoded success payload.</summary>
    private static byte[] Call(ITransport transport, DomainEndpoint endpoint, string service, string op, byte[] payload)
    {
        var raw = transport.Dial(endpoint.TcpAddr);
        var hostname = ExtractHostname(endpoint.TcpAddr);
        using var tls = TlsPinning.ConnectPinned(raw, hostname, endpoint.Fingerprints);

        var request = new RpcEnvelope.Request(service, op, payload);
        StreamFraming.SendFrame(tls, request.Encode());
        var respBytes = StreamFraming.ReadFrame(tls);
        var resp = RpcEnvelope.DecodeResponse(respBytes);

        if (!resp.IsOk)
        {
            throw new SdkException(resp.StatusCode, resp.Error ?? "unknown error");
        }

        return resp.Payload;
    }

    /// <summary>
    /// Fetch <paramref name="domain"/>'s currently-trusted public keys:
    /// <c>DomainKeys/get-domain-keys</c> over TCP CSIL-RPC, pinned to the domain's DNS
    /// <c>fp=</c> set, with signing keys pinned directly and encryption keys trusted
    /// only via a pinned signing key's vouch. <b>Always</b> also fetches
    /// <c>DomainKeys/get-revocations</c> for the same domain — regardless of what the
    /// <c>get-domain-keys</c> response's <c>RecentRevocationsAvailable</c> flag says — and
    /// drops any key a quorum-verified sibling revocation certificate targets.
    /// <c>RecentRevocationsAvailable</c> is an optional performance hint a well-behaved
    /// IDP may use to signal "you don't even need to ask"; a compromised/malicious or
    /// merely buggy IDP could otherwise use its absence to suppress this SDK from ever
    /// learning about a revocation, which is exactly the scenario revocation exists to
    /// guard against — so this SDK never uses it to skip the check. A
    /// <c>get-revocations</c> RPC or decode error is FATAL: this SDK must fail closed
    /// rather than silently proceed with a possibly-stale key set an attacker could have
    /// engineered by making the endpoint fail. An empty revocation list is normal
    /// success (nothing to apply). An empty trusted result (after applying revocations)
    /// is a fail-closed <see cref="SdkException"/>, matching the server's own posture.
    /// </summary>
    public static List<DomainPublicKey> FetchDomainKeys(ITransport transport, IDnsResolver dns, string domain)
    {
        var endpoint = DiscoverDomainEndpoint(dns, domain);

        var payload = Codec.EncodeEmptyRequest(new EmptyRequest());
        var respBytes = Call(transport, endpoint, "DomainKeys", "get-domain-keys", payload);
        var resp = Codec.DecodeGetDomainKeysResponse(respBytes);

        var trusted = new List<DomainPublicKey>(LinkKeys.LocalRp.Dns.Dns.TrustKeys(resp.Keys, endpoint.Fingerprints));
        if (trusted.Count == 0)
        {
            throw new SdkException(SdkException.ErrorKind.NoTrustedDomainKeys, domain);
        }

        // Always fetch revocations — never gated on RecentRevocationsAvailable (see this
        // method's doc comment). A failure here is FATAL: it must never be swallowed to
        // "just proceed unfiltered". Re-resolved (rather than reusing `endpoint`) so this
        // is a fresh, independently-pinned dial like every other RPC call this SDK makes
        // — matching `RedeemClaimTicket`'s own resolve-then-dial shape rather than
        // threading connection state across two logically separate calls.
        var revocationsEndpoint = DiscoverDomainEndpoint(dns, domain);
        var since = DateTimeOffset.UtcNow.AddDays(-30).ToString("O");
        var reqPayload = Codec.EncodeGetRevocationsRequest(new GetRevocationsRequest(since));
        GetRevocationsResponse revResp;
        try
        {
            var revRespBytes = Call(transport, revocationsEndpoint, "DomainKeys", "get-revocations", reqPayload);
            revResp = Codec.DecodeGetRevocationsResponse(revRespBytes);
        }
        catch (Exception e) when (e is not OutOfMemoryException and not StackOverflowException)
        {
            // Fail closed: an unreachable/erroring/malformed get-revocations call (RPC
            // transport failure, TLS failure, or a decode error) must never be silently
            // treated as "no revocations" — that would let a hostile or merely-unreliable
            // IDP suppress a key's revocation simply by dropping or corrupting this call.
            throw new SdkException(
                SdkException.ErrorKind.RevocationUnavailable, $"get-revocations failed for domain {domain}", e);
        }

        foreach (var cert in revResp.Revocations)
        {
            try
            {
                Revocation.VerifyRevocationCertificate(cert, trusted, domain);
                trusted.RemoveAll(k => k.KeyId == cert.TargetKeyId);
            }
            catch (RevocationError)
            {
                // certificate didn't meet quorum; key stays trusted
            }
        }

        if (trusted.Count == 0)
        {
            throw new SdkException(SdkException.ErrorKind.NoTrustedDomainKeys, domain);
        }

        return trusted;
    }

    /// <summary>
    /// Redeem a claim ticket with <paramref name="domain"/>'s IDP:
    /// <c>LocalRp/redeem-claim-ticket</c> over TCP CSIL-RPC, pinned via the domain's DNS
    /// <c>fp=</c> set. Unauthenticated at the transport layer (no client cert) — the
    /// redemption request itself is signed with the local RP's signing key, which is the
    /// possession proof the server checks.
    /// </summary>
    public static LocalRpTicketRedemptionResponse RedeemClaimTicket(
        ITransport transport, IDnsResolver dns, string domain, SignedLocalRpTicketRedemptionRequest signedRequest)
    {
        var endpoint = DiscoverDomainEndpoint(dns, domain);
        var payload = Codec.EncodeSignedLocalRpTicketRedemptionRequest(signedRequest);
        var respBytes = Call(transport, endpoint, "LocalRp", "redeem-claim-ticket", payload);
        return Codec.DecodeLocalRpTicketRedemptionResponse(respBytes);
    }
}

File: sdks/local-rp/dart/lib/src/rpc/rpc_client.dart
// CSIL-RPC over the injected [Transport], TLS-pinned to a domain's DNS
// `fp=` records -- this SDK's only network surface, per the design doc's
// "Required Network Access": domain public keys, revocations, and
// claim-ticket redemption, all unauthenticated-TLS TCP CSIL-RPC calls.
library;

import 'dart:io';
import 'dart:typed_data';

import '../dns/dns.dart' as dnsproto;
import '../dns/dns_resolver.dart';
import '../errors.dart';
import '../revocation.dart' as revocation;
import '../wire/codec.dart';
import '../wire/types.dart';
import 'rpc_envelope.dart';
import 'stream_framing.dart';
import 'tls_pinning.dart' as tls;
import 'transport.dart';

/// A discovered endpoint for a domain: its pinned trust-anchor fingerprints
/// and its CSIL-RPC TCP address.
class DomainEndpoint {
  final List<String> fingerprints;
  final String tcpAddr;
  const DomainEndpoint(this.fingerprints, this.tcpAddr);
}

/// Look up a domain's trust anchor + TCP endpoint over DNS TXT. Fails
/// closed: a missing/unparseable record, or a `_linkkeys` record with no
/// `fp=` entries, or a `_linkkeys_apis` record with no `tcp=` entry, is an
/// error.
Future<DomainEndpoint> discoverDomainEndpoint(
    DnsResolver dns, String domain) async {
  final anchorName = dnsproto.linkkeysDnsName(domain);
  final anchorTxts = await dns.txtLookup(anchorName);
  List<String>? fingerprints;
  for (final txt in anchorTxts) {
    try {
      final rec = dnsproto.parseLinkKeysTxt(txt);
      if (rec.fingerprints.isNotEmpty) {
        fingerprints = rec.fingerprints;
        break;
      }
    } on DnsParseError {
      // try the next TXT record
    }
  }
  if (fingerprints == null) {
    throw SdkException(SdkExceptionKind.dns,
        'no usable $anchorName TXT record with fp= entries');
  }

  final apisName = dnsproto.linkkeysApisDnsName(domain);
  final apisTxts = await dns.txtLookup(apisName);
  String? tcpAddr;
  for (final txt in apisTxts) {
    try {
      final apis = dnsproto.parseLinkKeysApisTxt(txt);
      if (apis.tcp != null) {
        tcpAddr = apis.tcp;
        break;
      }
    } on DnsParseError {
      // try the next TXT record
    }
  }
  if (tcpAddr == null) {
    throw SdkException(
        SdkExceptionKind.dns, 'no usable $apisName TXT record with tcp= entry');
  }

  return DomainEndpoint(fingerprints, tcpAddr);
}

String _extractHostname(String hostPort) {
  if (hostPort.startsWith('[')) {
    final end = hostPort.indexOf(']');
    if (end != -1) return hostPort.substring(1, end);
  }
  final idx = hostPort.lastIndexOf(':');
  return idx == -1 ? hostPort : hostPort.substring(0, idx);
}

/// The shape of [call]: dial + speak one CSIL-RPC request/response over
/// [endpoint]. Exposed as a typedef solely so tests can override HOW a call
/// is transported -- see [call]'s docs for why this exists and its (tight)
/// bounds.
typedef RpcCaller = Future<Uint8List> Function(Transport transport,
    DomainEndpoint endpoint, String service, String op, Uint8List payload);

/// Send one CSIL-RPC request over a fresh TLS connection to [endpoint] and
/// return the decoded success payload. This is the ONLY call path
/// production code (`complete.dart`) uses; [fetchDomainKeys] and
/// [redeemClaimTicket] accept an optional [RpcCaller] override purely as a
/// **test seam**.
///
/// That override exists for one documented reason: `dart:io`'s TLS stack
/// (BoringSSL) refuses to negotiate a TLS handshake at all when the server
/// presents an Ed25519 certificate (`NO_COMMON_SIGNATURE_ALGORITHMS`,
/// verified empirically against an `openssl`-minted Ed25519 cert before
/// writing this comment) -- unlike the JDK, which the Java reference SDK's
/// flow test relies on to serve a real Ed25519-certificate TLS server in
/// the same process. Since this protocol's TLS pinning REQUIRES an Ed25519
/// server certificate by construction (the pinned fingerprint is a
/// domain's Ed25519 signing key), no in-process flow test can exercise the
/// real TLS handshake path in Dart today. Flow tests therefore use an
/// [RpcCaller] that dials a real TCP socket and speaks real CSIL-RPC
/// framing to a real in-process fake IDP, skipping only the TLS handshake
/// step -- every other production code path (CBOR wire codec, the full
/// `LocalRp`/`Claims`/`Revocation` verification chain, DNS TXT parsing) is
/// exercised unchanged. The TLS pin-check logic itself
/// (`tls_pinning.dart`'s certificate DER parsing and fingerprint
/// comparison) is unit-tested separately, directly against
/// `openssl`-minted Ed25519 certificate bytes, since it cannot be
/// exercised through a live handshake. See the package README's "Known
/// limitations" section.
Future<Uint8List> call(Transport transport, DomainEndpoint endpoint,
    String service, String op, Uint8List payload) async {
  final raw = await transport.dial(endpoint.tcpAddr);
  final hostname = _extractHostname(endpoint.tcpAddr);
  SecureSocket? secure;
  try {
    secure = await tls.connectPinned(raw, hostname, endpoint.fingerprints);
    final request = RpcRequest(service: service, op: op, payload: payload);
    await sendFrame(secure, request.encode());
    final reader = FrameReader(secure);
    final respBytes = await reader.readFrame();
    final resp = RpcResponse.decode(respBytes);

    if (!resp.isOk) {
      throw SdkException.server(resp.status, resp.error ?? 'unknown error');
    }
    return resp.payload;
  } on SdkException {
    rethrow;
  } catch (e) {
    throw SdkException(SdkExceptionKind.transport, '$e', cause: e);
  } finally {
    secure?.destroy();
  }
}

/// Fetch [domain]'s currently-trusted public keys:
/// `DomainKeys/get-domain-keys` over TCP CSIL-RPC, pinned to the domain's
/// DNS `fp=` set, with signing keys pinned directly and encryption keys
/// trusted only via a pinned signing key's vouch. Always also fetches
/// `DomainKeys/get-revocations` for the same domain -- regardless of what
/// the `get-domain-keys` response's `recentRevocationsAvailable` flag says
/// -- and drops any key a quorum-verified sibling revocation certificate
/// targets. `recentRevocationsAvailable` is an optional performance hint a
/// well-behaved IDP may use to signal "you don't even need to ask"; a
/// compromised/malicious or merely buggy IDP could otherwise use its
/// absence to suppress this SDK from ever learning about a revocation,
/// which is exactly the scenario revocation exists to guard against -- so
/// this SDK never uses it to skip the check. A `get-revocations` RPC error
/// (connection failure or response decode failure) is FATAL: this SDK must
/// fail closed rather than silently proceed with a possibly-stale key set
/// an attacker could have engineered by making the endpoint fail. An empty
/// revocation list is normal success (nothing to apply). An empty trusted
/// result (before or after applying revocations) is a fail-closed
/// [SdkException].
Future<List<DomainPublicKey>> fetchDomainKeys(
    Transport transport, DnsResolver dns, String domain,
    {RpcCaller? caller}) async {
  final doCall = caller ?? call;
  final endpoint = await discoverDomainEndpoint(dns, domain);

  final payload = Codec.encodeEmptyRequest(const EmptyRequest());
  final respBytes = await doCall(
      transport, endpoint, 'DomainKeys', 'get-domain-keys', payload);
  final resp = Codec.decodeGetDomainKeysResponse(respBytes);

  var trusted = await dnsproto.trustKeys(resp.keys, endpoint.fingerprints);
  if (trusted.isEmpty) {
    throw SdkException(SdkExceptionKind.noTrustedDomainKeys, domain);
  }

  // Always fetch revocations -- never gated on `recentRevocationsAvailable`
  // (see this function's doc comment). A failure here propagates, i.e. is
  // FATAL: it must never be swallowed to "just proceed unfiltered".
  final since = DateTime.now()
      .toUtc()
      .subtract(const Duration(days: 30))
      .toIso8601String();
  final reqPayload =
      Codec.encodeGetRevocationsRequest(GetRevocationsRequest(since: since));
  final revRespBytes = await doCall(
      transport, endpoint, 'DomainKeys', 'get-revocations', reqPayload);
  final revResp = Codec.decodeGetRevocationsResponse(revRespBytes);
  for (final cert in revResp.revocations) {
    try {
      await revocation.verifyRevocationCertificate(cert, trusted, domain);
      trusted = trusted.where((k) => k.keyId != cert.targetKeyId).toList();
    } catch (_) {
      // certificate didn't meet quorum; key stays trusted
    }
  }

  if (trusted.isEmpty) {
    throw SdkException(SdkExceptionKind.noTrustedDomainKeys, domain);
  }
  return trusted;
}

/// Redeem a claim ticket with [domain]'s IDP:
/// `LocalRp/redeem-claim-ticket` over TCP CSIL-RPC, pinned via the domain's
/// DNS `fp=` set. Unauthenticated at the transport layer (no client cert)
/// -- the redemption request itself is signed with the local RP's signing
/// key, which is the possession proof the server checks.
Future<LocalRpTicketRedemptionResponse> redeemClaimTicket(
    Transport transport,
    DnsResolver dns,
    String domain,
    SignedLocalRpTicketRedemptionRequest signedRequest,
    {RpcCaller? caller}) async {
  final doCall = caller ?? call;
  final endpoint = await discoverDomainEndpoint(dns, domain);
  final payload =
      Codec.encodeSignedLocalRpTicketRedemptionRequest(signedRequest);
  final respBytes = await doCall(
      transport, endpoint, 'LocalRp', 'redeem-claim-ticket', payload);
  return Codec.decodeLocalRpTicketRedemptionResponse(respBytes);
}

/// A [RpcCaller] that dials a real TCP socket and speaks real CSIL-RPC
/// stream framing, but skips the TLS handshake entirely (no pinning, no
/// encryption). **Test-only.** See [call]'s docs for exactly why this
/// exists and what it does and does not cover.
Future<Uint8List> insecureCallForTesting(
    Transport transport,
    DomainEndpoint endpoint,
    String service,
    String op,
    Uint8List payload) async {
  final raw = await transport.dial(endpoint.tcpAddr);
  try {
    final request = RpcRequest(service: service, op: op, payload: payload);
    await sendFrame(raw, request.encode());
    final reader = FrameReader(raw);
    final respBytes = await reader.readFrame();
    final resp = RpcResponse.decode(respBytes);
    if (!resp.isOk) {
      throw SdkException.server(resp.status, resp.error ?? 'unknown error');
    }
    return resp.payload;
  } on SdkException {
    rethrow;
  } catch (e) {
    throw SdkException(SdkExceptionKind.transport, '$e', cause: e);
  } finally {
    raw.destroy();
  }
}
